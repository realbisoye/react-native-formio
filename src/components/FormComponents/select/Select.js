import React from 'react';
import SelectComponent from '../sharedComponents/Select';
import formiojs from '../../../formio';
import {Text} from 'react-native';
import {interpolate, serialize, raw} from '../../../util';
import get from 'lodash/get';

export default class Select extends SelectComponent {
  constructor(props) {
    super(props);
    this.getValueField = this.getValueField.bind(this);
    this.refreshItems = this.refreshItems.bind(this);
    this.loadMoreItems = this.loadMoreItems.bind(this);
    this.setResult = this.setResult.bind(this);
    this.getValueDisplay = this.getValueDisplay.bind(this);
  }

  componentDidMount() {
    switch (this.props.component.dataSrc) {
      case 'values':
        this.internalFilter = true;
        this.setState({
          selectItems: this.props.component.data.values
        });
        break;
      case 'json':
        try {
          if (typeof this.props.component.data.json === 'string') {
            this.items = JSON.parse(this.props.component.data.json);
          }
          else if (typeof this.props.component.data.json === 'object') {
            this.items = this.props.component.data.json;
          }
          else {
            this.items = [];
          }
        }
        catch (error) {
          this.items = [];
        }
        this.options.params = {
          limit: parseInt(this.props.component.limit) || 20,
          skip: 0
        };
        this.refreshItems = (input, url, append) => {
          // If they typed in a search, reset skip.
          if ((this.lastInput || input) && this.lastInput !== input) {
            this.lastInput = input;
            this.options.params.skip = 0;
          }
          let items = this.items;
          if (input) {
            items = items.filter(item => {
              // Get the visible string from the interpolated item.
              const value = interpolate(this.props.component.template, {item}).replace(/<(?:.|\n)*?>/gm, '');
              switch (this.props.component.filter) {
                case 'startsWith':
                  return value.toLowerCase().lastIndexOf(input.toLowerCase(), 0) === 0;
                case 'contains':
                default:
                  return value.toLowerCase().indexOf(input.toLowerCase()) !== -1;
              }
            });
          }
          items = items.slice(this.options.params.skip, this.options.params.skip + this.options.params.limit);
          this.setResult(items, append);
        };
        this.refreshItems();
        break;
      case 'custom':
        this.refreshItems = () => {
          try {
            /* eslint-disable no-unused-vars */
            const {data, row} = this.props;
            /* eslint-enable no-unused-vars */
            let selectItems = eval('(function(data, row) { var values = [];' + this.props.component.data.custom.toString() + '; return values; })(data, row)');
            if (!Array.isArray(selectItems)) {
              throw 'Didn\'t return an array.';
            }
            this.setState({
              selectItems
            });
          }
          catch (error) {
            this.setState({
              selectItems: []
            });
          }
        };
        this.refreshItems();
        break;
      case 'resource':
      case 'url':
        if (this.props.component.dataSrc === 'url') {
          this.url = this.props.component.data.url;
          if (this.url.substr(0, 1) === '/') {
            this.url = formiojs.getBaseUrl() + this.props.component.data.url;
          }

          // Disable auth for outgoing requests.
          if (!this.props.component.authenticate && this.url.indexOf(formiojs.getBaseUrl()) === -1) {
            this.options = {
              disableJWT: true,
              headers: {
                Authorization: undefined,
                Pragma: undefined,
                'Cache-Control': undefined
              }
            };
          }
        }
        else {
          this.url = formiojs.getBaseUrl();
          if (this.props.component.data.project) {
            this.url += '/project/' + this.props.component.data.project;
          }
          this.url += '/form/'  + this.props.component.data.resource + '/submission';
        }

        this.options.params = {
          limit: this.props.component.limit || 100,
          skip: 0
        };

        this.refreshItems = (input, newUrl, append) => {
          let {data, row} = this.props;
          newUrl = newUrl || this.url;
          // Allow templating the url.
          newUrl = interpolate(newUrl, {
            data,
            row,
            formioBase: formiojs.getBaseUrl()
          });
          if (!newUrl) {
            return;
          }

          // If this is a search, then add that to the filter.
          if (this.props.component.searchField && input) {
            // If they typed in a search, reset skip.
            if (this.lastInput !== input) {
              this.lastInput = input;
              this.options.params.skip = 0;
            }
            newUrl += ((newUrl.indexOf('?') === -1) ? '?' : '&') +
              encodeURIComponent(this.props.component.searchField) +
              '=' +
              encodeURIComponent(input);
          }

          // Add the other filter.
          if (this.props.component.filter) {
            const filter = interpolate(this.props.component.filter, {data});
            newUrl += ((newUrl.indexOf('?') === -1) ? '?' : '&') + filter;
          }

          // If they wish to return only some fields.
          if (this.props.component.selectFields) {
            this.options.params.select = this.props.component.selectFields;
          }

          // If this is a search, then add that to the filter.
          newUrl += ((newUrl.indexOf('?') === -1) ? '?' : '&') + serialize(this.options.params);
          formiojs.request(newUrl).then(data => {
            // If the selectValue prop is defined, use it.
            if (this.props.component.selectValues) {
              this.setResult(get(data, this.props.component.selectValues, []), append);
            }
            // Attempt to default to the formio settings for a resource.
            else if (data.hasOwnProperty('data')) {
              this.setResult(data.data, append);
            }
            else if (data.hasOwnProperty('items')) {
              this.setResult(data.items, append);
            }
            // Use the data itself.
            else {
              this.setResult(data, append);
            }
          });
        };

        this.refreshItems();

        break;
      default:
        this.setState({
          selectItems: []
        });
    }
  }

  getValueField() {
    if (this.props.component.dataSrc === 'custom' || this.props.component.dataSrc === 'json') {
      return false;
    }
    if (this.props.component.dataSrc === 'resource' && this.props.component.valueProperty === '') {
      return '_id';
    }
    return this.props.component.valueProperty || 'value';
  }

  refreshItems() {}

  loadMoreItems(event) {
    event.stopPropagation();
    event.preventDefault();
    this.props.onEvent('loadMore', this.loadMoreItems);
    this.options.params.skip += this.options.params.limit;
    this.refreshItems(null, null, true);
  }

  setResult(data, append) {
    if (!Array.isArray(data)) {
      data = [data];
    }
    if (append) {
      this.setState({
        selectItems: [...this.state.selectItems, ...data],
        hasNextPage: this.state.selectItems.length >= (this.options.params.limit + this.options.params.skip),
      });
    }
 else {
      this.setState({
        selectItems: data,
        hasNextPage: this.state.selectItems.length >= (this.options.params.limit + this.options.params.skip),
      });
    }
  }

  getValueDisplay(component, data) {
    const getItem = (data) => {
      switch (component.dataSrc) {
        case 'values':
          component.data.values.forEach((item) => {
            if (item.value === data) {
              data = item;
            }
          });
          return data;
        case 'json':
          if (component.valueProperty) {
            let selectItems;
            try {
              selectItems = JSON.parse(component.data.json);
            }
            catch (error) {
              selectItems = [];
            }
            selectItems.forEach((item) => {
              if (item[component.valueProperty] === data) {
                data = item;
              }
            });
          }
          return data;
        // TODO: implement url and resource view.
        case 'url':
        case 'resource':
        default:
          return data;
      }
    };
    if (component.multiple && Array.isArray(data)) {
      return data.map(getItem).reduce(function(prev, item) {
        var value;
        if (typeof item === 'object') {
          value = (<Text>{raw(interpolate(component.template, {item}))}</Text>);
        }
        else {
          value = item;
        }
        return (prev === '' ? '' : ', ') + value;
      }, '');
    }
    else {
      var item = getItem(data);
      var value;
      if (typeof item === 'object') {
        value = (<Text>{raw(interpolate(component.template, {item}))}</Text>);
      }
      else {
        value = item;
      }
      return value;
    }
  }
}
