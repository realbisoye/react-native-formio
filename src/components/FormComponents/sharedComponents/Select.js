import React from 'react';
import {View, Text, StyleSheet} from 'react-native';
import Icon from 'react-native-vector-icons/dist/FontAwesome';
import {Dropdown} from 'react-native-material-dropdown';
import MultiSelect from 'react-native-multiple-select';
import {FormLabel} from 'react-native-elements';
import get from 'lodash/get';
import isEqual from 'lodash/isEqual';
import PropTypes from 'prop-types';
import DeviceInfo from 'react-native-device-info';
import ValueComponent from './Value';
import Tooltip from './Tooltip';
import colors from '../../../defaultTheme/colors';

export default class SelectComponent extends ValueComponent {
  constructor(props) {
    super(props);
    this.data = {...this.props.data};
    this.state = {
      selectItems: [],
      searchTerm: '',
      hasNextPage: false,
      open: false
    };
    this.willReceiveProps = this.willReceiveProps.bind(this);
    this.onChangeSelect = this.onChangeSelect.bind(this);
    this.onSearch = this.onSearch.bind(this);
    this.valueField = this.valueField.bind(this);
    this.getElements = this.getElements.bind(this);
    this.onToggle = this.onToggle.bind(this);
  }

  willReceiveProps(nextProps) {
    if (this.props.component.refreshOn && !nextProps.formPristine) {
      const refreshOn = this.props.component.refreshOn;
      this.refresh = false;
      if (refreshOn === 'data') {
        if (!isEqual(this.data, nextProps.data)) {
          this.refresh = true;
        }
      }
      else {
        if ((!this.data.hasOwnProperty(refreshOn) && nextProps.hasOwnProperty(refreshOn)) || this.data[refreshOn] !== nextProps.data[refreshOn]) {
          this.refresh = true;
        }
        else if (nextProps && nextProps.row && nextProps.row.hasOwnProperty(refreshOn) && this.props.row[refreshOn] !== nextProps.row[refreshOn]) {
          this.refresh = true;
        }
      }
      if (this.refresh && this.props.component.clearOnRefresh) {
        this.setValue(this.getDefaultValue());
      }
    }
    if (this.refresh) {
      this.refreshItems();
      this.refresh = false;
    }
    this.data = {...nextProps.data};
  }

  valueField() {
    let valueFieldItem = this.props.component.valueProperty || 'value';
    if (typeof this.getValueField === 'function') {
      valueFieldItem = this.getValueField();
    }
    return valueFieldItem;
  }

  onChangeSelect(selected) {
    let value;
    if (this.props.component.multiple) {
      value = this.state.selectItems.filter((i) => selected.includes(i.label))
      .map((i) => i.value);
    }
 else {
      value = selected;
    }
    if (Array.isArray(value) && this.valueField()) {
      value.forEach((val, index) => {
        value[index] = (typeof val === 'object' ? get(val, this.valueField()) : val);
      });
    }
    else if (typeof value === 'object' && this.valueField()) {
      value = get(value, this.valueField());
    }
    this.setValue(value);
  }

  onSearch(text) {
    this.setState({
      searchTerm: text
    });
    if (typeof this.refreshItems === 'function') {
      this.refreshItems(text);
    }
  }

  elementLayout(position) {
    switch (position) {
      case 'top':
       return {
          flexDirection: 'column',
        };
      case 'left-left':
      case 'left-right':
        return {
          flexDirection: 'row',
          alignItems: 'flex-start',
        };
      case 'right-left':
      case 'right-right':
        return {
          flexDirection: 'row-reverse',
          flex: 1,
          marginHorizontal: 20,
        };
      case 'bottom':
        return {
          flexDirection: 'column-reverse',
        };
      default:
        return {
          flexDirection: 'column',
        };
    }
  }

  onToggle(isOpen) {
    this.props.onEvent('selectToggle', this, isOpen);
    this.setState(prevState => {
      prevState.open = isOpen;
      return prevState;
    });
  }

  getElements() {
    const selectStyle = StyleSheet.create({
      wrapper: {
        flex: 1,
        marginHorizontal: 20,
        marginTop: 20
      },
      container: {
        zIndex: 1000,
      },
      containerSingle: {
        flex: 1,
        marginTop: 0,
        marginBottom: 15,
      },
      label: {
        maxWidth: DeviceInfo.isTablet() ? 580 : 210,
        color: this.props.theme.Label.color,
        fontSize: DeviceInfo.isTablet() ? this.props.theme.Label.fontSize : 12,
      },
      mainElement: this.elementLayout(this.props.component.labelPosition),
      labelWrapper: {
        flexDirection: 'row',
        marginTop: this.props.component.labelPosition === 'top' || this.props.component.labelPosition === 'bottom' ? 0 : 40,
        marginRight: this.props.component.labelPosition === 'left-left' || this.props.component.labelPosition === 'left-right' ? 10 : 0,
      },
      list: {
        backgroundColor: colors.mainBackground,
      },
      descriptionText: {
        fontSize: DeviceInfo.isTablet() ? 12 : 10,
        marginLeft: 20,
        marginTop: 10,
      },
    });

    const {component} = this.props;
    const labelText = component.label && !component.hideLabel ? component.label : '';
    const requiredInline = (!component.label && component.validate && component.validate.required ? <Icon name='asterisk' /> : '');
    const multiMode = component.multiple;
    let values;
    let Element;

    const inputLabel = labelText ?
      <FormLabel labelStyle={selectStyle.label}>{requiredInline} {component.label}</FormLabel> : null;

    if (multiMode) {
      values = this.state.value && this.state.value.item ? this.state.selectItems.filter((i) => this.state.value.item.includes(i.value)) : [];
      Element = (
        <MultiSelect
          hideTags
          hideSubmitButton
          fixedHeight
          items={this.state.selectItems}
          uniqueKey={'label'}
          displayKey={'label'}
          selectText={component.placeholder}
          showDropDowns={true}
          readOnlyHeadings={true}
          onSelectedItemsChange={this.onChangeSelect}
          selectedItems={values.map(v => v.label)}
          tagRemoveIconColor={this.props.colors.primary1Color}
          selectedItemTextColor={this.props.colors.primary1Color}
          selectedItemIconColor={this.props.colors.primary1Color}
        />
      );
    }
    else {
      values = this.state.value ? this.state.value.item : '';
      Element = (
        <Dropdown
          label={component.placeholder}
          data={this.state.selectItems}
          disabled={this.props.readOnly}
          style={selectStyle.label}
          dropdownOffset={{top: 40, left: 5}}
          containerStyle={selectStyle.containerSingle}
          itemTextStyle={selectStyle.listText}
          value={values}
          onChangeText={this.onChangeSelect}
          labelFontSize={18}
        />
      );
    }
    return (
      <View style={selectStyle.wrapper}>
        <View style={selectStyle.mainElement}>
          <View style={selectStyle.labelWrapper}>
          {inputLabel}
          {component.tooltip && <Tooltip
            text={component.tooltip}
            color={this.props.colors.alternateTextColor}
            backgroundColor={this.props.colors.primary1Color}
            styles={{
              icon: {
                marginTop: this.props.component.labelPosition === 'top' || this.props.component.labelPosition === 'bottom' ? 0 : 12,
                marginLeft: this.props.component.labelPosition === 'top' || this.props.component.labelPosition === 'bottom' ? -15 : -20,
              }
            }}
          />}
          </View>
          {Element}
        </View>
        {component.description && <Text style={selectStyle.descriptionText}>{component.description}</Text>}
      </View>
    );
  }
}

SelectComponent.propTypes = {
  data: PropTypes.any,
  component: PropTypes.any,
  row: PropTypes.object,
  readOnly: PropTypes.bool,
  value: PropTypes.any,
  onEvent: PropTypes.func,
};
