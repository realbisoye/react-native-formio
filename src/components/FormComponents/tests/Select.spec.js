import React from 'react';
import renderer from 'react-test-renderer';
import colors from '../../../defaultTheme/colors';
import theme from '../../../defaultTheme';
import Select from '../select/Select';

describe('Select Field', () => {
  describe('Select Component', () => {
    var component= {
      'input': true,
      'tableView': true,
      'label': '',
      'key': 'select',
      'placeholder': '',
      'data': {
        'values': [
          {
            'value': 'a',
            'label': 'a'
          },
          {
            'value': 'b',
            'label': 'b'
          },
          {
            'value': 'c',
            'label': 'c'
          },
          {
            'value': 'd',
            'label': 'd'
          }
        ],
        'json': '',
        'url': '',
        'resource': ''
      },
      'dataSrc': 'values',
      'valueProperty': '',
      'defaultValue': '',
      'refreshOn': '',
      'filter': '',
      'authenticate': false,
      'template': '<span>{{ item.label }}</span>',
      'multiple': false,
      'protected': true,
      'unique': false,
      'persistent': true,
      'validate': {
        'required': false
      },
      'type': 'select',
      'conditional': {
        'show': '',
        'when': null,
        'eq': ''
      }
    };
    var attachToForm = jest.fn();

    it.only('Render a basic select component', () => {
      const element = renderer.create(
        <Select
          component={component}
          colors={colors}
          theme={theme}
          attachToForm={attachToForm}
        />
      );
      expect(element).toMatchSnapshot();
    });

    it('Check the label of the select component', () => {
      component.label = 'Select component';
      const element = renderer.create(
        <Select
          component={component}
          attachToForm={attachToForm}
        />
      );
      expect(element.html()).to.equal(component.label);
      expect(element).to.have.length(1);
    });

    it('Check with out label for select component', () => {
      component.label = null;
      const element = renderer.create(
        <Select
          component={component}
          attachToForm={attachToForm}
        />
      );
      expect(element.html()).to.equal(component.label);
      expect(element).to.have.length(0);
    });

    it('Check the dropdown and check the number of options of select component', () => {
      const element = renderer.create(
        <Select
          component={component}
          attachToForm={attachToForm}
        />
      );
      expect(element.find('.rw-open')).to.have.length(0);
      element.find('.rw-i-caret-down').simulate('click');
      expect(element.find('.rw-open')).to.have.length(1);
      expect(element.find('.rw-popup')).to.have.length(1);

      //To check the number of options for select component

      expect(element.find('.rw-popup ul li').length).to.equal(component.data.values.length);
    });

    it('Render a multiple select component ', () => {
      component.multiple = true;
      const element = renderer.create(
        <Select
          component={component}
          attachToForm={attachToForm}
        />
      );
      expect(element).to.have.length(1);
      component.multiple = false;
    });
  });
});
