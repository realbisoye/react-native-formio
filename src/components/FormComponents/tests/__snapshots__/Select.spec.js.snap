// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Field Select Component Render a basic select component 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "marginHorizontal": 20,
      "marginTop": 20,
    }
  }
>
  Dropdown
</View>
`;
